scripts/diffinc_restore.sh

```diff
@@ -0,0 +1,139 @@
+fi
+if [ ! -f "${IOTSTACK_DIR}/menu.sh" ]; then
+  echo "./menu.sh file was not found. $IOTSTACK_DIR does not seem to be an IOTstack directory."
+  exit 1
+fi
+
+#check backup file, set if not specified
+if [ -z "${1+x}" ]; then
+  #not set, find in default directory
+  BACKUP_DIR="$IOTSTACK_DIR/backups/diffincbackup"
+  PRESENT_BACKUPS=($(find "$BACKUP_DIR" -maxdepth 1 -name "backup_base-????-??-??_????_*_*.tar*"))
+  if [[ ! -z "$PRESENT_BACKUPS" ]]; then
+    #found backup files, find the most recent base backup
+    PRESENT_BACKUPS=("${PRESENT_BACKUPS[@]##*/}")
+    readarray -t PRESENT_BACKUPS_SORTED < <(printf '%s\0' "${PRESENT_BACKUPS[@]}" | sort -z | xargs -0n1)
+    SAMPLE_FILE="${PRESENT_BACKUPS_SORTED[-1]}"
+  else
+    echo "No backup file provided and no suitable backup file found in $IOTSTACK_DIR/backups/diffinc_backup"
+    exit 1
+  fi
+  
+else
+  if [ ! -f "$1" ]; then
+    echo "ERROR: supplied file $1 does not exist."
+    exit 1
+  fi
+  BACKUP_DIR=${1%/*}
+  SAMPLE_FILE="${1##*/}"
+  if [[ $SAMPLE_FILE != backup_base-????-??-??_????_*_*.tar* ]]; then
+    echo "ERROR: filename of supplied file $SAMPLE_FILENAME is not the correct format."
+    exit 1
+  fi
+fi
+
+#a sample file is found, find and check all associated backups
+IFS='._' read -r -a FILENAME_ARR <<< "$SAMPLE_FILE" #splits filename in array delimited by . and _
+BACKUP_TIME="${FILENAME_ARR[1]}_${FILENAME_ARR[2]}"
+BACKUP_FILE_SET=($(find "$BACKUP_DIR" -maxdepth 1 -name "backup_${BACKUP_TIME}_*_*.tar*"))
+#sort to get highest level and reinstate in order
+readarray -t BACKUP_FILE_SET_SORTED < <(printf '%s\0' "${BACKUP_FILE_SET[@]##*/}" | sort -z | xargs -0n1)
+for BACKUP in ${BACKUP_FILE_SET_SORTED[@]}; do 
+  IFS='._' read -ra FILENAME_ARR <<< "$BACKUP" #splits filename in array delimited by . and _
+  SERVICES_MULTI+=(${FILENAME_ARR[3]}) #extracts main/nextcloud ect
+  LEVELS+=(${FILENAME_ARR[4]}) #extracts backup levels
+done
+IFS=" " read -r -a SERVICES <<< "$(tr ' ' '\n' <<< "${SERVICES_MULTI[@]}" | sort -u | tr '\n' ' ')" #will contain main
+EXPECTED_FILECOUNT=$(( (${#SERVICES[@]}) * (${LEVELS[-1]} + 1) ))
+if [ $EXPECTED_FILECOUNT -ne ${#BACKUP_FILE_SET_SORTED[@]} ]; then
+  echo "#################################"
+  echo "WARNING! Unexpected number of backup files for base backup $SAMPLE_FILE. Number of services: ${#SERVICES[@]}, highest level: "${LEVELS[0]}", therefore expected $EXPECTED_FILECOUNT files but found ${#PRESENT_BACKUPS_SORTED[@]}. "
+  echo "Please check the backups! Found files:"
+  printf '%s\n' "${BACKUP_FILE_SET_SORTED[@]}"
+  echo "#################################"
+else
+  echo "Found ${EXPECTED_FILECOUNT} files associated with the backup, files are:"
+  printf '%s\n' "${BACKUP_FILE_SET_SORTED[@]}"
+fi
+
+#backup validated as far as possible, confirm user intent
+echo "Do you wish to continue  restoring from backup? This will delete the present data! (y/n)"
```

Comment:
Maybe we should move the current data to a `.old` directory?

Reply:
Implemented it as a user selectable option in [cd86f9c](https://github.com/SensorsIot/IOTstack/pull/640/commits/cd86f9cfa65ee2e34a2cb8e3d536d5bc12c0137b) because it may become an issue when dealing with limited disk space.